docs/content/docs/dev/table/data_stream_api.md

```diff
@@ -0,0 +1,2106 @@
+--------------------------------------------
+
+The DataStream API uses instances of `org.apache.flink.api.common.typeinfo.TypeInformation` to describe
+the record type that travels in the stream. In particular, it defines how to serialize and deserialize
+records from one DataStream operator to the other. It also helps in serializing state into savepoint
+and checkpoints.
+
+The Table API uses custom data structures to represent records internally and exposes `org.apache.flink.table.types.DataType`
+to users for declaring the external format into which the data structures are converted for easier
+usage in sources, sinks, UDFs, or DataStream API.
+
+`DataType` is richer than `TypeInformation` as it also includes details about the logical SQL type.
+Therefore, some details will be added implicitly during the conversion.
+
+Column names and types of a `Table` are automatically derived from the `TypeInformation` of the
+`DataStream`. Use `DataStream.getType()` to check whether the type information has been detected
+correctly via the DataStream API's reflective type extraction facilities. If the outermost record's
+`TypeInformation` is a `CompositeType`, it will be flattened in the first level when deriving a table's
+schema.
+
+### TypeInformation to DataType
+
+The following rules apply when converting `TypeInformation` to a `DataType`:
+
+- All subclasses of `TypeInformation` are mapped to logical types including nullability that is aligned
+with Flink's built-in serializers.
+
+- Subclasses of `TupleTypeInfoBase` are translated into a row (for `Row`) or structured type (for tuples,
+POJOs, and case classes).
+
+- `BigDecimal` is converted to `DECIMAL(38, 18)` by default.
+
+- The order of `PojoTypeInfo` fields is determined during the conversion by looking for a constructor
+that orders all fields with its parameters. Otherwise the field order will be alphabetical.
+
+- `GenericTypeInfo` and other `TypeInformation` that cannot be represented as one of the listed
+`org.apache.flink.table.api.DataTypes` will be treated as a black-box `RAW` type. The current session
+configuration is used to materialize the serializer of the raw type. Composite nested fields will not
+be accessible then.
+
+- See {{< gh_link file="flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/utils/TypeInfoDataTypeConverter.java" name="TypeInfoDataTypeConverter" >}} for the full translation logic.
+
+Use `DataTypes.of(TypeInformation)` to call the above logic in custom schema declaration or in UDFs.
+
+### DataType to TypeInformation
+
+The table runtime will make sure to properly serialize the output records to the first operator of the
+DataStream API.
+
+{{< hint warning >}}
+Afterwards, the type information semantics of the DataStream API need to be considered.
+{{< /hint >}}
+
+{{< top >}}
+
+Legacy Conversion
+-----------------
+
+{{< hint info >}}
+The following section describes outdated parts of the API that will be removed in future versions.
```

Comment:
If this is outdated and the methods are declared `@Deprecated`, I would vote for actually removing this section for the sake of concise documentation that doesn't confuse users. If this is only relevant to previous Flink versions, it could live in their branches' docs, but shouldn't be carried on.

Reply:
Unfortunately, the methods are not marked as `@Deprecated` yet. Due to time constraints, this feature made it last minute and is fully usable with 1.13.1. I would suggest to keep the docs around for at least another release.

Reply:
We clarified offline: unfortunately, these APIs are not deprecated yet and the new ones are still experimental, so we'll have to keep this section in for now.